### ISSUE TITLE
API soundness issues

### ISSUE THREAD

**Issue by parasyte** [parasyte/pixels#238]
This was mentioned to me yesterday by @Lokathor. Trying to recreate a `Pixels` at runtime (e.g. in response to an event) causes validation errors and crashes.

I found an ugly workaround using `Option<Pixels>` and calling `take()` on it to drop the internal references before trying to recreate it. To me, this suggests the bug is related to https://github.com/gfx-rs/wgpu/issues/1463 (and this very long thread: https://github.com/rust-windowing/raw-window-handle/issues/71). The conversation is a bit hard to follow, but I think I can summarize it as: `RawWindowHandle` needs to be treated like a raw pointer that references the `HasRawWindowHandle` implementer. In this case, `Window` is the implementor, and `Surface` holds a `RawWindowHandle` pointing to it. This means that their lifetimes are required to be bound together, but there is nothing in the API contract enforcing that. It's left up to the caller of unsafe surface creation: https://github.com/parasyte/pixels/blob/4c6682e65bb417cdaa6e9964cbace5369149452d/src/builder.rs#L220

Whoops.

Pixels holds ownership of a surface, and it looks like having two surfaces alive pointing to the same `RawWindowHandle` is bad news bears. (This is just a best guess.)

Similarly, our API allows dropping the window after creating `Pixels`, and that is very clearly unsound. We can't borrow the window, since that would prevent moving it later (even though it's perfectly fine to do so). Almost every user moves their window into the event-loop closure to call `window.request_redraw()` on it, at a minimum.

I think the best way to solve this on our end is to take ownership of the window. That would prevent the issue as reported (can't recreate `Pixels` without creating a new `Window`) and it also solves this other related issue (can't drop `Window` without dropping `Pixels`). I don't know if shared ownership (with `Rc` or `Arc`) is acceptable, since that would probably just result in validation errors and crashes again if multiple surfaces are created from one window.

**Comment by ZetaNumbers** [comment]
So we should mark `SurfaceTexture::new` as unsafe for now, right?

**Comment by Lokathor** [comment]
That's a breaking change that's quick and dirty. If you're going to issue a breaking change you probably are better served to fix the whole problem in one go.

**Comment by parasyte** [comment]
I do agree with the last comment. It seems like it should not be much more work to take ownership of the window. The advantage is that it would keep the API safe.

I would love to see a solution for https://github.com/gfx-rs/wgpu/issues/1463 that we could take immediate advantage of but seems like it won't be available soon. @ZetaNumbers if you need a fix for this bug, I can expedite the work on the window ownership proposal. I'm also open to other suggestions.
